### PR TITLE
[FIX JENKINS-39625] Page preloading with pipeline state - Runs

### PR DESCRIPTION
**Pull request by tfennelly**
# Description

See [JENKINS-39625](https://issues.jenkins-ci.org/browse/JENKINS-39625).

This is following on from #612 (which is closed now - all changes on this PR now and pointing to `master`), adding `PipelineStatePreloader` extension point impl for preloading pipeline pages with state that can be used to short-circuit the fetch layer.

What `PipelineStatePreloader` does is adds an entry in the `$blueocean.prefetchdata.PipelineStatePreloader` page object on page creation, with that object containing the `restUrl` for the first page of activity data i.e. records `0` to `26`. The client-side `Fetch` code (`blueocen-core-js`) checks the `restUrl` values in the objects in `$blueocean.prefetchdata` before issuing the REST API call. If the URL matches the fetch url then the fetch call bypasses the fetch and uses the `data` value inside the `$blueocean.prefetchdata.PipelineStatePreloader` object.

screenshot 2016-12-13 16 15 26

We should be able to do this for other page data too e.g. favorites.

# Coding Todos:
- [x] Generic commons code for parsing a Blue Ocean UI URL. See 	ee8773d and 1712012.
- [x] Implement `PipelineStatePreloader` for setting pipeline runs state on the page. See 5bd676b.
- [x] Modify client-side fetch code to check the "prefetch" data on the loaded page. We can attach prefetch data and associate it with REST API calls, allowing the fetch code to bypass the actual fetch call and use the prefetched data.

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [x] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [x] Ran Acceptance Test Harness against PR changes. https://ci.blueocean.io/job/ATH-Jenkinsfile/job/master/616/

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

@jenkinsci/code-reviewers @reviewbybees

### CODE REVIEW

**Comment by imeredith** [comment]
@tfennelly now that the mobx stores are in, it will be a bit different. If its a page with paged data, we will need a way to create the Pager class against that data. Will not be hard. Ditto for activities

**Comment by tfennelly** [comment]
@imeredith yeah, once that stuff is in mobx then we will need to send a different "store update instruction". I haven't gotten to that part yet anyway - see the last unchecked item in the "Code Todos" checklist.

**Comment by imeredith** [comment]
Nice way of implementing it on the javascript side!! I was thinking about how to inject it into stores but this is really nice because of how generic it is.

File: blueocean-commons/src/main/java/io/jenkins/blueocean/commons/BlueoceanUrl.java [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+                                            // e.g. /blue/organizations/jenkins/f1%2Ff3%20with%20spaces%2Ff3%20pipeline/detail/magic-branch-X/...
+                                            blueoceanUrl.addPart(UrlPart.BRANCH, urlDecode(urlTokens.next()));
+                                            if (urlTokens.hasNext()) {
+                                                // e.g. /blue/organizations/jenkins/f1%2Ff3%20with%20spaces%2Ff3%20pipeline/detail/magic-branch-X/55/...
+                                                blueoceanUrl.addPart(UrlPart.PIPELINE_RUN_DETAIL_ID, urlDecode(urlTokens.next()));
+                                                if (urlTokens.hasNext()) {
+                                                    next = urlTokens.next();
+                                                    if (PIPELINE_RUN_DETAIL_TABS.contains(next.toLowerCase())) {
+                                                        // e.g. /blue/organizations/jenkins/f1%2Ff3%20with%20spaces%2Ff3%20pipeline/detail/magic-branch-X/55/pipeline
+                                                        blueoceanUrl.addPart(UrlPart.PIPELINE_RUN_DETAIL_TAB, next.toLowerCase());
```
Comment (michaelneale) [inline blueocean-commons/src/main/java/io/jenkins/blueocean/commons/BlueoceanUrl.java]: holy indentation batman!

File: blueocean-commons/src/test/java/io/jenkins/blueocean/commons/BlueoceanUrlTest.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.net.MalformedURLException;
+
+/**
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ */
+public class BlueoceanUrlTest {
```
Comment (michaelneale) [inline blueocean-commons/src/test/java/io/jenkins/blueocean/commons/BlueoceanUrlTest.java]: BlueOceanUrlTest *nit* - but it's test code so its fine ;)

**Comment by i386** [comment]
If I hit http://pr.sandbox.blueocean.io/pr-613/blue/organizations/jenkins/multibranch/detail/stephenc-patch-1/1/pipeline/ I can still see http://pr.sandbox.blueocean.io/pr-613/blue/rest/organizations/jenkins/pipelines/multibranch/branches/stephenc-patch-1/runs/1/ being loaded. Shouldn't that not be happening now?

**Comment by michaelneale** [comment]
nice one! seems to work well trying it here: http://pr.sandbox.blueocean.io/pr-613/blue/organizations/jenkins/multibranch/activity

File: blueocean-commons/src/main/java/io/jenkins/blueocean/commons/BlueoceanUrl.java [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+ * This class performs a "best effort" attempt to parse a URL as a Blue Ocean
+ * URL, extracting what it thinks are the relevant "parts" and making available via the
+ * {@link #getPart(UrlPart)} and {@link #hasPart(UrlPart)} functions.
+ * <p>
+ * See TBD comment on {@link UrlPart}.
+ *
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ */
+@Restricted(NoExternalUse.class) // Internal use only for now because there's a fair chance we'll change how it works. See TBD comment on UrlPart.
+public class BlueoceanUrl {
```
Comment (michaelneale) [inline blueocean-commons/src/main/java/io/jenkins/blueocean/commons/BlueoceanUrl.java]: nit `BlueOceanUrl`

File: blueocean-commons/src/main/java/io/jenkins/blueocean/commons/BlueoceanUrl.java (hunk 1 above)
Comment (michaelneale) [inline blueocean-commons/src/main/java/io/jenkins/blueocean/commons/BlueoceanUrl.java]: I expect there is probably some pattern to make this much neater.

File: blueocean-core-js/src/js/capability/CapabilityAugmenter.js [hunk 4]
```diff
@@ -3,7 +3,20 @@
  */
 
 const addClass = (clazz, classMap) => {
-    const className = clazz._class;
+    let className;
+
+    if (Array.isArray(clazz._class)) {
+        // If it's an array of class names, just take the first.
+        // TODO: Hmmm ... not sure if this is the right thing to do when we have an array of class names.
```
Comment (michaelneale) [inline blueocean-core-js/src/js/capability/CapabilityAugmenter.js]: Good Q - did you get to the bottom of this?

**Comment by michaelneale** [comment]
This has been around for a while, probably wouldn't hurt to update to master as well. How happy is the ATH with it?

**Comment by michaelneale** [comment]
Actually this up to date - not sure why github wasn't telling me that. Passes sniff test locally - although as @i386 pointed out, it seems to be fetching things when it shouldn't... so looks close.

**Comment by tfennelly** [comment]
@i386 Wrt your comment ... it's not happening on that page because that's a Run Details page and this PR just does it for the pipeline/job activity page. As I said in the description, my plan is to add prefetches for more data on different pages as part of follow-up PRs. The main thing in these PRs is to put the mechanics in place. After that, it should be relatively easy to add the same for others.

That said ... the specific one you point out should be easy to add since it's just a URL test in the existing `PipelineStatePreloader`.

**Comment by michaelneale** [comment]
@tfennelly @i386 ah - I thought james meant the activity page  - not results, in that case, carry on (they can be done as separate PRs if you like... this is probably big enough as it is).

File: blueocean-core-js/src/js/fetch.js [hunk 5]
```diff
@@ -5,6 +5,9 @@ import utils from './utils';
 import config from './config';
 import dedupe from './utils/dedupe-calls';
 import urlconfig from './urlconfig';
+import { prefetchdata } from './scopes';
+
+const Promise = es6Promise.Promise;
```
Comment (tfennelly) [inline blueocean-core-js/src/js/fetch.js]: Where? I'm not importing anything new. Just accessing the constructor.

**Comment by tfennelly** [comment]
@michaelneale @i386 In fairness, the JIRA does ask for the Run Details page. Not sure why I ended up doing it for the pipeline activity - duh. Anyway, same thing really ... these PRs provide the mechanism to now do mini PRs to prefectch other stuff, so I think we should just run with this for now.

File: blueocean-commons/src/main/java/io/jenkins/blueocean/commons/BlueoceanUrl.java (hunk 1 above)
Comment (michaelneale) [inline blueocean-commons/src/main/java/io/jenkins/blueocean/commons/BlueoceanUrl.java]: @tfennelly I am ashamed to admit that this is quite easy to read like this! - I could see quickly that it was deliving into the URL hierarchy - there is probably a list/map way of doing it that keeps that clear hierarchy though.

**Comment by michaelneale** [comment]
@tfennelly ok - fair enough. Good to have it for activity as that is already slow - so if we can have it with RunDetails page - awesome!

File: blueocean-core-js/src/js/fetch.js (hunk 5 above)
Comment (imeredith) [inline blueocean-core-js/src/js/fetch.js]: ah right. I don't think you need to do that? At the very least i have used the contructor in other places without it. But I don't thing it hurts also so meh.

**Comment by michaelneale** [comment]
LGTM from me  but would like a bee from @imeredith to cross check with quite a few changes.

File: blueocean-commons/src/main/java/io/jenkins/blueocean/commons/BlueoceanUrl.java (hunk 3 above)
Comment (imeredith) [inline blueocean-commons/src/main/java/io/jenkins/blueocean/commons/BlueoceanUrl.java]: I would like to see the refactor as part of pr

**Comment by imeredith** [comment]
I'm happy with this PR apart from casing on BlueoceanUrl. Would like to see it refactored before merging this pr otherwise it probably wont happen and then plugins will start using it down the line somewhere :) :bee: apart fom that though.

**Comment by tfennelly** [comment]
@imeredith `BlueoceanUrl` has been renamed to `BlueUrlTokenizer`.

**Comment by tfennelly** [comment]
FYI ... I will be recreating this PR as a new PR that has this and the "Part 1" PR, just to make it easier to merge.

File: blueocean-commons/src/main/java/io/jenkins/blueocean/commons/PageStatePreloader.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+ * Allows the loading page's JavaScript blueocean global scope to
+ * be pre-populated with data that we know the page is going to need, thereby
+ * providing a mechanism for eliminating the request overhead for that data.
+ *
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ */
+public abstract class PageStatePreloader extends ExtensionPoint {
+
+    /**
+     * Get the JavaScript object graph path at shiwh the state is to be stored.
```
Comment (scherler) [inline blueocean-commons/src/main/java/io/jenkins/blueocean/commons/PageStatePreloader.java]: shiwh ?

Comment (michaelneale) [inline blueocean-commons/src/main/java/io/jenkins/blueocean/commons/PageStatePreloader.java]: German for ...

**Comment by michaelneale** [comment]
@tfennelly some failures in the ATH look like they are picking up something legit: 

```
- Open acitivty page wait for first run to finish [1;33m(21.713s)[0m
2016-12-14 22:12:25,199 [Thread-2] INFO   (InputStreamHandler.java:47) com.github.eirslett.maven.plugins.frontend.lib.DefaultGulpRunner - [0;31m   Timed out while waiting for element <.activity-table tr#testResults-1 svg circle.unstable> to be present for 20000 milliseconds. [1;37m - expected [0;32m"found"[0m[0m but got: [0;31m"not found"[0m[0m
```

Also, as @imeredith mentioned - could the BlueoceanUrl be renamed correctly before others start building on this?

**Comment by michaelneale** [comment]
yeah sorry @tfennelly I can see that testResults.js error locally too occasionaly (not every second run, but intermittent) - it is most strange... @imeredith any ideas?

**Comment by michaelneale** [comment]
also this is missing the feature that was really requested in the main ticket, which was there for a reason... so perhaps needs more time.

**Comment by tfennelly** [comment]
@michaelneale @imeredith FYI, the ATH failures were fixed by https://github.com/jenkinsci/blueocean-acceptance-test/commit/b3b447e3389c4f127fadbd60032799bae073774f. Why is this happening now, I hear you ask .... classic ATH kind of flaky testing scenario ... the new code just pushed timings into a zone that made us hit the situation more easily.

**Comment by michaelneale** [comment]
nice !

File: blueocean-commons/src/main/java/io/jenkins/blueocean/commons/PageStatePreloader.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package io.jenkins.blueocean.commons;
+
+import hudson.ExtensionList;
+import org.apache.tools.ant.ExtensionPoint;
```
Comment (amuniz) [inline blueocean-commons/src/main/java/io/jenkins/blueocean/commons/PageStatePreloader.java]: Is this intentional? Shouldn't it be `hudson.ExtensionPoint`?

Comment (amuniz) [inline blueocean-commons/src/main/java/io/jenkins/blueocean/commons/PageStatePreloader.java]: cf. https://github.com/jenkinsci/blueocean-plugin/pull/1124

Comment (tfennelly) [inline blueocean-commons/src/main/java/io/jenkins/blueocean/commons/PageStatePreloader.java]: @amuniz sure does look wrong .... no idea how that actually works, but it does, from what I can remember.

**Pull request merged**